### ISSUE TITLE
Lazy streams: GC retains memory for too long due to premature promotion of objects to maxgen

### ISSUE THREAD

**Issue by minad** [cisco/ChezScheme#513]
While working with lazy streams I found that the Chez GC retains objects for too long leading to a temporary space leak. The memory is reclaimed after a while, but the memory usage is huge.
Expected behavior would be constant memory usage.

In this example, the streams are implemented as pairs with a delayed cdr. Since the head is strict, these are odd streams according to [Wadler et al.](https://www.researchgate.net/publication/2646969_How_to_Add_Laziness_to_a_Strict_Language_Without_Even_Being_Odd). In contrast, srfi-40, 41 and 45 discuss even streams where the head is also delayed. But it turns out even streams are prone to space leaks. 

The odd streams here should not lead to leaks since all the functions are tail recursive. In particular, the stream head `(count 0)` is not retained. There is no need to keep many elements of the stream alive.

~~~
(define (count n)
  (cons n (delay (count (add1 n)))))

(define (filter p s)
  (if (p (car s))
      (cons (car s) (delay (filter p (force (cdr s)))))
      (filter p (force (cdr s)))))

(define (print n s)
  (when (> n 0)
    (display (car s))
    (display "\n")
    (print (sub1 n) (force (cdr s)))))

(print 100 (filter (lambda (n) (> n 50000000)) (count 0)))
~~~

**Comment by jltaylor-us** [comment]
Can't do a whole lot with that code without the definitions for `delay` and `force`.

**Comment by jltaylor-us** [comment]
oops, premature comment... please ignore.

**Comment by jltaylor-us** [comment]
This "temporary space leak" behavior is inherent in using a generational garbage collector with data structures like this.  If you set the `collect-trip-bytes` parameter lower (so the garbage collector runs more often) then you'll see a corresponding reduction in memory used by this example.  (Alternately, if you were doing some interesting work on each item this would also allocate memory and cause the garbage collector to run more often.)

**Comment by minad** [comment]
@jltaylor-us Thank you for your response! Setting `collect-trip-bytes` to a lower value helps - the memory usage is constant as it should. It also works if I set `collect-maximum-generation` to a smaller value (effectively disabling the generational gc). But using a high `collect-trip-bytes` value combined with many generations leads to the memory explosion (hundreds of MiB resident memory, while `collect-trip-bytes` seems to be 8MiB by default). Could you please elaborate a bit as to why that happens? 

I would expect the intermediate thunks, which are only referenced for a short period of time, to not even survive the youngest generation. To state my expectations more precisely, I would expect constant memory usage proportional to the value of `collect-trip-bytes`. I wonder why there is a dependency on the number of generations.

**Comment by jltaylor-us** [comment]
Well, _something_ must be live when the GC runs in the middle of the program.  Whatever that is is promoted into next generation, and will remain there until a GC of that generation runs.  If the GC that triggered this promotion was already for generation 1 or 2 then the object will be around for a little while (even if it was freshly created).  The fact that a newly created object can skip straight into an older generation is mentioned briefly in [CSUG](http://cisco.github.io/ChezScheme/csug9.5/smgmt.html#./smgmt:h1):

> The storage manager places newly allocated objects into generation 0. During a generation 0 collection, objects in generation 0 that survive the collection move, by default, to generation 1. Similarly, during a generation 1 collection, objects in generations 0 and 1 that survive move to generation 2, and so on. During the collection of the maximum nonstatic collection, all surviving nonstatic objects move (possibly back) into the maximum nonstatic generation. With this mechanism, it is possible for an object to skip one or more generations, but this is not likely to happen to many objects, and if the objects become inaccessible, their storage is reclaimed eventually. 

So, perhaps I was not quite correct in saying that this behavior is inherent to generational garbage collectors in general, but I believe it is inherent for any that have the above property.

Based on my quick reading about the V8 garbage collector, the reason none of the stream elements make it out of its "young generation" is that it collects the young generation into itself, but maintains additional information so that it can tell which objects have survived more than one of these collections and promote them into the next generation.  The fact that a stream element is live during one of these collections doesn't promote it to the next generation unless it was _also_ live during the previous collection.

As a side note, I vaguely remember that Chez tries to gracefully handle being denied more memory by the OS, so everything might work just fine if you ran with a memory limit.  I also remember that last time I tried doing such a thing on MacOS the ulimit setting was cheerfully ignored by whatever part of the OS is supposed to enforce it, so I can't try it out at the moment.

**Comment by minad** [comment]
> Well, something must be live when the GC runs in the middle of the program.

I checked what this is by printing the object counts, and as expected the retained objects are mostly pairs and procedures (aka the stream elements) as expected.

>>  During the collection of the maximum nonstatic collection, all surviving nonstatic objects move (possibly back) into the maximum nonstatic generation.

Thank you! This is the crucial detail - the current stream head is copied to the maximum generation and then memory reclamation happens only when the maximum generation is collected the next time, and then memory grows again heavily and we get this sawtooth like pattern. And the next collection of the maximum generation won't happen for quite a while due to the fixed gc-roundtrip based scheduling.

> maintains additional information so that it can tell which objects have survived more than one of these collections and promote them into the next generation.

Yes, this would be the right thing to do. I've written myself a few garbage collectors in different variants, among them one copying with multiple generations and remembered sets. Therefore this Chez behavior went against my experience. Maybe the Chez GC could be fixed to handle such scenarios more gracefully. I think the above program is a particularly bad edge case exposing the undesired behavior you cited above:

>> During the collection of the maximum nonstatic collection, all surviving nonstatic objects move (possibly back) into the maximum nonstatic generation. With this mechanism, it is possible for an object to skip one or more generations, **but this is not likely to happen to many objects**, and if the objects become inaccessible, their storage is reclaimed eventually.

The edge case behavior is even worsened due to the gc-roundtrip based selection of the generation to be collected. A smarter (and also less deterministic) mechanism could probably mitigate the issue too, admittedly introducing its own sets of issues.

But clearly, this is an isolated edge case which would probably not occur as such in a real world program. But then I think a GC should handle the trivial edge cases well enough in order to handle the more complex real scenarios.

> As a side note, I vaguely remember that Chez tries to gracefully handle being denied more memory by the OS, so everything might work just fine if you ran with a memory limit.

On my Linux system the program works correctly in a sense of not crashing and terminating. However while doing that it uses almost all memory for a while. And then it nicely returns everything to the OS after the maximum collection. But using up all memory has its own huge cost since the program is basically page faulting (and during the fault, entering the kernel) all the time. In contrast, in the ideal scenario, where the thunks do not survive the nursery the whole nursery memory always fit into the cpu cache, allowing much much faster allocation, copying etc.

In the default gc settings, the above program runs in 23s on my notebook. If I set `collect-maximum-generation` to 1, the program terminates instead in 6s.

**Comment by minad** [comment]
For reference, it seems someone else observed this toowhile running chez. https://stackoverflow.com/questions/59214946/why-does-my-lazy-filtered-list-in-scheme-consume-so-much-memory

**Comment by gwatt** [comment]
Part of the answer is that delay does more work than I think you think it does. Look at the definition of [delay](https://github.com/cisco/ChezScheme/blob/master/s/syntax.ss#L7978) in terms of [make-promise](https://github.com/cisco/ChezScheme/blob/6ad42f74d052861dcbb51204002587421662215b/s/4.ss#L388)

You'll see that scheme promises handle multiple values, recursive promises, and memoize the result. Your example code makes use of none of that functionality.

Out of curiosity, what was the similar javascript program? If I replace `(delay expr)` with `(lambda () expr)` The memory usage goes way down and the speed goes way up.

```scheme
; Using standard delay
(time (print 100 (filter (lambda (n) (> n 50000000)) (count 0))))
    570 collections
    26.340055624s elapsed cpu time, including 24.415856711s collecting
    26.360049577s elapsed real time, including 24.436703107s collecting
    4800499504 bytes allocated, including 4315926192 bytes reclaimed

; Using (lambda () expr) instead of (delay expr)
(time (print 100 (filter (lambda (n) (> n 50000000)) (count 0))))
    190 collections
    1.644209020s elapsed cpu time, including 0.957043907s collecting
    1.645823428s elapsed real time, including 0.958372374s collecting
    1600243312 bytes allocated, including 2578932192 bytes reclaimed
```

However, if you have expensive operations that are generating your tails and need the memoization, delay will be better than just using lambda

**Comment by minad** [comment]
@gwatt Yes, you are right. Sharing is not used here and I suspected that the complexity of delay makes it harder for the compiler and optimizer to see what is going on. (This is also the reason why I didn't originally attribute this issue to the GC).

Such code as the above, which does not use sharing, but is formulated with sharing constructs, might occur if you are using a stream library. In my case the issue occured when I was working on a compiler backend with one target being Chez. Maybe it makes sense to check if I could somehow use cheaper lazy values, but I assumed delay to not doing unnecessary work. I have to take look.

Nonetheless the GC behavior in this case is not optimal.

Are there plans to improve upon the existing GC? Since right now the GC is copying, maybe also considering latency given large heaps. I think I've seen something about incremental gc come up before in the racket on chez project.

**Comment by minad** [comment]
The js implementation used objects as thunks `{forced:false,fn:function{...}}`. I will try what happens if I switch the Chez implementation to use structs instead of closures to represent thunks. But I don't expect much gain here.

**Comment by dybvig** [comment]
The problem is that lazy lists (streams) are built up from head to tail via side effects.  (In this case, the side effects occur when promises are forced.)  This leads to a series of pointers from older objects to younger objects, violating the basic assumption of generational collectors that such pointers are uncommon.  At some point, the current tail of a lazy list may be promoted to an older generation, and this pins the entire rest of the list (even the parts not yet allocated) in memory until the next collection of that generation.

This wouldn't be so bad, except, as @jltaylor-us points out, Chez Scheme's collector can promote some objects undeservedly to older generations, including even freshly allocated objects to the oldest nonstatic generation, because it always promotes all accessible objects to the target generation.  This was a poor design decision when I made it, so a fix is long overdue (about three decades).  The fix, of course, is to modify the collector to move objects up only one generation at a time.  Anyone interested in attempting this might want to let me know; otherwise, I might take it on myself.  The task is nontrivial, and the benefits are probably limited for most programs, which is why it has remained the way it is for so long.

In the meantime, some limited user-level solutions to this problem are available.  As @gwatt observed, replacing promises with thunks eliminates the problem.  Of course, it also causes a stream traversed more than once to be built more than once.  And as @minad observed, one can alleviate the problem by modifying collection parameters, e.g., to use fewer generations or collect older generations more frequently, though of course this might be an unfortunate choice in general.  A program can also explicitly truncate a lazy list by clearing tail pointers, say when the program has some way to know that only a few elements of the list need to be retained, such as when the list is used for some fixed amount of lookahead.  Or a program might use weak pairs to avoid retaining stream tails across garbage collections while effectively caching stream values between collections.

**Comment by minad** [comment]
@dybvig Thank you! Good to hear that you are considering improving the behavior of the generational GC. What was the reason for the original design decision, to perform promotion of all objects to the target generation? Simpler GC implementation or faster evacuation performance, since you only have to copy to one target memory area?

**Comment by dybvig** [comment]
I've just created pull request #518, which modifies the collector to promote one generation at a time.  On my mac laptop, a build of the master branch runs the code above in just under 9s in about 2G of memory, measured via `(maximum-memory-bytes)`.   With the modification,  it takes just under 3s and maxes out at about 54M, which is only a few MB above the maximum memory bytes at startup.

**Comment by dybvig** [comment]
@minad, I don't remember what my thinking was at the time, but the cost of relocating an object does go up with incremental promotion because of the need to track pointers from older newspace objects to younger newspace objects.  For example, say a generation 1 object points to a generation 0 object.  If both are promoted to generation 2, there's no longer a pointer from an older object to a younger object. If, however, the generation 0 object is promoted to generation 1 and the generation 1 object is promoted to generation 0, the collector needs to record that fact when relocating the objects.  Indeed, the collector ran much slower on my mac with my first cut.  Most of this turned out to be because I had recoded the relocate macros as forceinline functions, and llvm's compiler on the mac did a poor job with the inlining.  (gcc on my linux machine didn't have this problem.)  So a reverted back to the macros.  I eliminated the rest of the performance penalty by a set of other changes, the most important of which was to run a specialized collector (compiled from the same source code) when collecting generation 0 into generation 1 with no generation 0 locked objects, which typically covers 75% of collections with the default collector settings in code that doesn't make heavy use of locked objects.  In the end, my benchmark runs showed a slight net benefit from the pull request changes of around 1-2%.
